File: tablet_qt/tasks/icd10mixed.cpp
/*
    Copyright (C) 2012-2018 Rudolf Cardinal (rudolf@pobox.com).

    This file is part of CamCOPS.

    CamCOPS is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    CamCOPS is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with CamCOPS. If not, see <http://www.gnu.org/licenses/>.
*/

#include "icd10mixed.h"
#include "common/appstrings.h"
#include "common/textconst.h"
#include "lib/datetime.h"
#include "maths/mathfunc.h"
#include "lib/stringfunc.h"
#include "lib/uifunc.h"
#include "questionnairelib/commonoptions.h"
#include "questionnairelib/namevaluepair.h"
#include "questionnairelib/qudatetime.h"
#include "questionnairelib/questionnaire.h"
#include "questionnairelib/quheading.h"
#include "questionnairelib/qumcqgrid.h"
#include "questionnairelib/qutext.h"
#include "questionnairelib/qutextedit.h"
#include "tasklib/taskfactory.h"
using datetime::shortDate;
using mathfunc::allTrue;
using mathfunc::anyFalse;
using stringfunc::standardResult;
using uifunc::trueFalseUnknown;

const QString Icd10Mixed::ICD10MIXED_TABLENAME("icd10mixed");

const QString DATE_PERTAINS_TO("date_pertains_to");
const QString COMMENTS("comments");
const QString MIXTURE_OR_RAPID_ALTERNATION("mixture_or_rapid_alternation");
const QString DURATION_AT_LEAST_2_WEEKS("duration_at_least_2_weeks");
const QStringList CRITERIA{
    MIXTURE_OR_RAPID_ALTERNATION,
    DURATION_AT_LEAST_2_WEEKS,
};

void initializeIcd10Mixed(TaskFactory& factory)
{
    static TaskRegistrar<Icd10Mixed> registered(factory);
}


Icd10Mixed::Icd10Mixed(CamcopsApp& app, DatabaseManager& db,
                       const int load_pk) :
    Task(app, db, ICD10MIXED_TABLENAME, false, true, false)  // ... anon, clin, resp
{
    addField(DATE_PERTAINS_TO, QVariant::Date);
    addField(COMMENTS, QVariant::String);
    addField(MIXTURE_OR_RAPID_ALTERNATION, QVariant::Bool);
    addField(DURATION_AT_LEAST_2_WEEKS, QVariant::Bool);

    load(load_pk);  // MUST ALWAYS CALL from derived Task constructor.

    // Extra initialization:
    if (load_pk == dbconst::NONEXISTENT_PK) {
        setValue(DATE_PERTAINS_TO, datetime::nowDate(), false);
    }
}


// ============================================================================
// Class info
// ============================================================================

QString Icd10Mixed::shortname() const
{
    return "ICD10-mixed";
}


QString Icd10Mixed::longname() const
{
    return tr("ICD-10 symptomatic criteria for a mixed affective episode "
              "(as in e.g. F06.3, F25, F38.00, F31.6)");
}


QString Icd10Mixed::menusubtitle() const
{
    return textconst::ICD10;
}


QString Icd10Mixed::infoFilenameStem() const
{
    return "icd";
}


// ============================================================================
// Instance info
// ============================================================================

bool Icd10Mixed::isComplete() const
{
    return !meetsCriteria().isNull();
}


QStringList Icd10Mixed::summary() const
{
    return QStringList{
        standardResult(appstring(appstrings::DATE_PERTAINS_TO),
                       shortDate(value(DATE_PERTAINS_TO))),
        standardResult(textconst::MEETS_CRITERIA,
                       trueFalseUnknown(meetsCriteria())),
    };
}


QStringList Icd10Mixed::detail() const
{
    QStringList lines = completenessInfo() + summary() + QStringList{
        fieldSummary(COMMENTS, textconst::EXAMINER_COMMENTS),
        fieldSummary(MIXTURE_OR_RAPID_ALTERNATION, xstring("a")),
        fieldSummary(DURATION_AT_LEAST_2_WEEKS, xstring("b")),
    };
    return lines;
}


OpenableWidget* Icd10Mixed::editor(const bool read_only)
{
    const NameValueOptions true_false_options = CommonOptions::falseTrueBoolean();
    QVector<QuestionWithOneField> qfields{
        {fieldRef(MIXTURE_OR_RAPID_ALTERNATION), xstring("a")},
        {fieldRef(DURATION_AT_LEAST_2_WEEKS), xstring("b")},
    };

    QuPagePtr page((new QuPage{
        getClinicianQuestionnaireBlockRawPointer(),
        (new QuText(appstring(appstrings::ICD10_SYMPTOMATIC_DISCLAIMER)))->setBold(),
        new QuText(appstring(appstrings::DATE_PERTAINS_TO)),
        (new QuDateTime(fieldRef(DATE_PERTAINS_TO)))
            ->setMode(QuDateTime::Mode::DefaultDate)
            ->setOfferNowButton(true),
        new QuMcqGrid(qfields, true_false_options),
        new QuHeading(textconst::COMMENTS),
        new QuTextEdit(fieldRef(COMMENTS, false)),
    })->setTitle(longname()));

    Questionnaire* questionnaire = new Questionnaire(m_app, {page});
    questionnaire->setType(QuPage::PageType::Clinician);
    questionnaire->setReadOnly(read_only);
    return questionnaire;
}


// ============================================================================
// Task-specific calculations
// ============================================================================

QVariant Icd10Mixed::meetsCriteria() const
{
    const QVector<QVariant> criteria = values(CRITERIA);
    if (allTrue(criteria)) {
        return true;
    }
    if (anyFalse(criteria)) {
        return false;
    }
    return QVariant();  // don't know yet
}

File: server/camcops_server/tasks/icd10mixed.py
#!/usr/bin/env python
# camcops_server/tasks/icd10mixed.py

"""
===============================================================================

    Copyright (C) 2012-2018 Rudolf Cardinal (rudolf@pobox.com).

    This file is part of CamCOPS.

    CamCOPS is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    CamCOPS is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with CamCOPS. If not, see <http://www.gnu.org/licenses/>.

===============================================================================
"""

from typing import List, Optional

from cardinal_pythonlib.datetimefunc import format_datetime
from cardinal_pythonlib.typetests import is_false
import cardinal_pythonlib.rnc_web as ws
from sqlalchemy.sql.schema import Column
from sqlalchemy.sql.sqltypes import Boolean, Date, UnicodeText

from camcops_server.cc_modules.cc_constants import (
    CssClass,
    DateFormat,
    ICD10_COPYRIGHT_DIV,
)
from camcops_server.cc_modules.cc_ctvinfo import CTV_INCOMPLETE, CtvInfo
from camcops_server.cc_modules.cc_html import (
    get_true_false_none,
    tr_qa,
)
from camcops_server.cc_modules.cc_request import CamcopsRequest
from camcops_server.cc_modules.cc_sqla_coltypes import (
    BIT_CHECKER,
    CamcopsColumn,
)
from camcops_server.cc_modules.cc_summaryelement import SummaryElement
from camcops_server.cc_modules.cc_task import (
    Task,
    TaskHasClinicianMixin,
    TaskHasPatientMixin,
)


# =============================================================================
# Icd10Mixed
# =============================================================================

class Icd10Mixed(TaskHasClinicianMixin, TaskHasPatientMixin, Task):
    __tablename__ = "icd10mixed"
    shortname = "ICD10-MIXED"
    longname = (
        "ICD-10 symptomatic criteria for a mixed affective episode "
        "(as in e.g. F06.3, F25, F38.00, F31.6)"
    )

    date_pertains_to = Column(
        "date_pertains_to", Date,
        comment="Date the assessment pertains to"
    )
    comments = Column(
        "comments", UnicodeText,
        comment="Clinician's comments"
    )
    mixture_or_rapid_alternation = CamcopsColumn(
        "mixture_or_rapid_alternation", Boolean,
        permitted_value_checker=BIT_CHECKER,
        comment="The episode is characterized by either a mixture or "
                "a rapid alternation (i.e. within a few hours) of hypomanic, "
                "manic and depressive symptoms."
    )
    duration_at_least_2_weeks = CamcopsColumn(
        "duration_at_least_2_weeks", Boolean,
        permitted_value_checker=BIT_CHECKER,
        comment="Both manic and depressive symptoms must be prominent"
                " most of the time during a period of at least two weeks."
    )

    def get_clinical_text(self, req: CamcopsRequest) -> List[CtvInfo]:
        if not self.is_complete():
            return CTV_INCOMPLETE
        category = (
            ("Meets" if self.meets_criteria() else "Does not meet") +
            " criteria for mixed affective episode"
        )
        infolist = [CtvInfo(
            content="Pertains to: {}. {}.".format(
                format_datetime(self.date_pertains_to, DateFormat.LONG_DATE),
                category
            )
        )]
        if self.comments:
            infolist.append(CtvInfo(content=ws.webify(self.comments)))
        return infolist

    def get_summaries(self, req: CamcopsRequest) -> List[SummaryElement]:
        return self.standard_task_summary_fields() + [
            SummaryElement(
                name="meets_criteria",
                coltype=Boolean(),
                value=self.meets_criteria(),
                comment="Meets criteria for a mixed affective episode?"),
        ]

    # Meets criteria? These also return null for unknown.
    def meets_criteria(self) -> Optional[bool]:
        if (self.mixture_or_rapid_alternation and
                self.duration_at_least_2_weeks):
            return True
        if is_false(self.mixture_or_rapid_alternation):
            return False
        if is_false(self.duration_at_least_2_weeks):
            return False
        return None

    def is_complete(self) -> bool:
        return (
            self.meets_criteria() is not None and
            self.field_contents_valid()
        )

    def get_task_html(self, req: CamcopsRequest) -> str:
        h = """
            {clinician_comments}
            <div class="{CssClass.SUMMARY}">
                <table class="{CssClass.SUMMARY}">
                    {tr_is_complete}
                    {date_pertains_to}
                    {meets_criteria}
                </table>
            </div>
            <div class="{CssClass.EXPLANATION}">
                {icd10_symptomatic_disclaimer}
            </div>
            <table class="{CssClass.TASKDETAIL}">
                <tr>
                    <th width="80%">Question</th>
                    <th width="20%">Answer</th>
                </tr>
                {mixture_or_rapid_alternation}
                {duration_at_least_2_weeks}
            </table>
            {ICD10_COPYRIGHT_DIV}
        """.format(
            clinician_comments=self.get_standard_clinician_comments_block(
                req, self.comments),
            CssClass=CssClass,
            tr_is_complete=self.get_is_complete_tr(req),
            date_pertains_to=tr_qa(
                req.wappstring("date_pertains_to"),
                format_datetime(self.date_pertains_to, DateFormat.LONG_DATE,
                                default=None)
            ),
            meets_criteria=tr_qa(
                req.wappstring("meets_criteria"),
                get_true_false_none(req, self.meets_criteria())
            ),
            icd10_symptomatic_disclaimer=req.wappstring(
                "icd10_symptomatic_disclaimer"),
            mixture_or_rapid_alternation=self.get_twocol_bool_row_true_false(
                req, "mixture_or_rapid_alternation", self.wxstring(req, "a")),
            duration_at_least_2_weeks=self.get_twocol_bool_row_true_false(
                req, "duration_at_least_2_weeks", self.wxstring(req, "b")),
            ICD10_COPYRIGHT_DIV=ICD10_COPYRIGHT_DIV,
        )
        return h
